### PR TITLE
Update ucx to 1.12.0; add variants

### PR DESCRIPTION
**Pull request by iarspider**

### CODE REVIEW

File: var/spack/repos/builtin/packages/ucx/package.py [hunk 1]
```diff
@@ -19,6 +20,8 @@ class Ucx(AutotoolsPackage, CudaPackage):
     # Current
     version('1.12.1', sha256='40b447c8e7da94a253f2828001b2d76021eb4ad39647107d433d62d61e18ae8e')
     version('1.12.0', sha256='93e994de2d1a4df32381ea92ba4c98a249010d1720eb0f6110dc72c9a7d25db6')
+    version('1.12.1', sha256='40b447c8e7da94a253f2828001b2d76021eb4ad39647107d433d62d61e18ae8e')
```
Comment (hppritcha) [inline var/spack/repos/builtin/packages/ucx/package.py]: why are these extra 1.12.1 and 1.12.0 lines added?

File: var/spack/repos/builtin/packages/ucx/package.py [hunk 2]
```diff
@@ -127,11 +157,8 @@ def configure_args(self):
...
 
-        # Activate SIMD based on properties of the target
-        if 'avx' in self.spec.target:
-            config_args.append('--with-avx')
-        else:
-            config_args.append('--without-avx')
+        rdmac_prefix = lambda x: self.spec['rdma-core'].prefix \
+            if 'rdma-core' in self.spec else None
 
         config_args.extend(self.enable_or_disable('optimizations'))
```
Comment (hppritcha) [inline var/spack/repos/builtin/packages/ucx/package.py:161]: did you mean to use the enable-compiler-opt configure option here?  --enable-optimizations is different

**Review by hppritcha — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/ucx/package.py (hunk 1 above)
Comment (iarspider) [reply var/spack/repos/builtin/packages/ucx/package.py]: Whoops. Copy-paste error.

File: var/spack/repos/builtin/packages/ucx/package.py (hunk 2 above)
Comment (iarspider) [reply var/spack/repos/builtin/packages/ucx/package.py:161]: I have added handling of `opt` variant (i.e. `--enable-compiler-opt=foo`)

**Review by hppritcha — APPROVED** [review]

**Review by hppritcha — APPROVED** [review]

**Review by tldahlgren — APPROVED** [review]

File: var/spack/repos/builtin/packages/ucx/package.py [hunk 3]
```diff
@@ -51,6 +54,8 @@ class Ucx(AutotoolsPackage, CudaPackage):
             description='Enable logging')
     variant('debug', default=False,
             description='Enable debugging')
+    variant('opt', default='0', values=('0', '1', '2', '3'), multi=False,
```
Comment (haampie) [inline var/spack/repos/builtin/packages/ucx/package.py:57]: why 0?

File: var/spack/repos/builtin/packages/ucx/package.py [hunk 4]
```diff
@@ -127,13 +155,12 @@ def configure_args(self):
...
-        if 'avx' in self.spec.target:
-            config_args.append('--with-avx')
-        else:
-            config_args.append('--without-avx')
+        rdmac_prefix = lambda x: self.spec['rdma-core'].prefix \
+            if 'rdma-core' in self.spec else None
 
         config_args.extend(self.enable_or_disable('optimizations'))
+        config_args.append('--enable-compiler-opt=' +
+                           self.spec.variant['opt'].value)
```
Comment (haampie) [inline var/spack/repos/builtin/packages/ucx/package.py:163]: did you try this? it's a typo.

File: var/spack/repos/builtin/packages/ucx/package.py (hunk 3 above)
Comment (iarspider) [reply var/spack/repos/builtin/packages/ucx/package.py:57]: Yup, did not read the code correctly. It should be 3.

**Pull request merged**